### PR TITLE
install: Use UResult

### PR DESCRIPTION
**Pull request by Funky185540**
Related to uutils#2464

I didn't actually check whether the error messages match what GNUs `install` gives, I merely took the existing ones. I hope I got the idea right, otherwise I'm open for suggestions.

### CODE REVIEW

**Comment by tertsdiepraam** [comment]
Amazing! Great work with the custom error type!

> I didn't actually check whether the error messages match what GNUs install gives, I merely took the existing ones.

That's good enough for now :)

I have two suggestions:
1. Some of the variants of the custom error are essentially `io::Error`s, requiring manual conversion from that to the custom error. You could use `UIoError` instead. Most of the time, you take a function that returns `io::Result`, say `fs::metadata()` and use `map_err_context` to turn it into a `UResult`, which you can then return:
```rust
fn md(path: Path) -> UResult<Metadata> {
    fs::metadata(path).map_err_context(|| format!("could not get metadata for {}", path)) 
}
```
See the docs for more information: https://docs.rs/uucore/0.0.9/uucore/error/struct.UIoError.html (let me know if you have any questions about this, I'd also love feedback on improving the docs). This will also give error messages that are more in line with GNU.
2. There's an opportunity to remove the `all_successful` variables by using the `show!` method. Docs are admittedly lacking a bit in this department, but there's the end of the [usage section](https://docs.rs/uucore/0.0.9/uucore/error/index.html#usage) and the docs for [`set_exit_code`](https://docs.rs/uucore/0.0.9/uucore/error/fn.set_exit_code.html) and `touch` uses this technique as well (in `uumain`).

**Comment by Funky185540** [comment]
> Some of the variants of the custom error are essentially io::Errors, requiring manual conversion from that to the custom error. You could use UIoError instead. Most of the time, you take a function that returns io::Result, say fs::metadata() and use map_err_context to turn it into a UResult, which you can then return:

I saw that while attempting to understand how the UError type works. What I particularly like about the custom error types is that all error messages are in one place for me to edit if the need arises. 

But right now I can't think of a way how to mangle the regular std::io::Error into the UIoError (in an elegant way) such that I can use the `Display`-trait implementation that's already there. Hm...
Is this something that could be solved with a macro? I haven't worked with macros yet, so I'm quite unaware of their potential.

> There's an opportunity to remove the all_successful variables by using the show! method. Docs are admittedly lacking a bit in this department, but there's the end of the usage section and the docs for set_exit_code and touch uses this technique as well (in uumain).

Oh right, I read about `set_exit_code` and asked myself while implementing "When would you need that when you have custom error types?". *That* of course makes a lot of sense. I'll update the code!

**Comment by Funky185540** [comment]
The code now makes use of `show!` for noncritical errors that don't cause immediate program abortion.

I'll be honest: the sheer wealth of different errors is pretty overwhelming, so I didn't do any "manual" tests to see whether I broke some error outputs. I'm entirely relying on the tests for this...

**Comment by tertsdiepraam** [comment]
> What I particularly like about the custom error types is that all error messages are in one place for me to edit if the need arises.

I understand. However, I've noticed that most IO errors need some context, like "could not get metadata for ...", which describes the operation that is being done in that part of the code. So I think it makes sense to put those descriptions where they occur.

I wonder if there is some change we can make to the API to get the best of both worlds.

> But right now I can't think of a way how to mangle the regular std::io::Error into the UIoError (in an elegant way) such that I can use the Display-trait implementation that's already there.

I wouldn't prefer this over `UIoError`, but we could take that `Display impl` and split it off into a separate function that takes an `io::ErrorKind` which you could then reuse.

**Comment by Funky185540** [comment]
> I understand. However, I've noticed that most IO errors need some context, like "could not get metadata for ...", which describes the operation that is being done in that part of the code. So I think it makes sense to put those descriptions where they occur.

Good point. But if multiple utils fail to get Metadata, isn't that a candidate for a `CommonError` then? Maybe one could "attach" (perhaps by means of another attribute that's a `Option<std::io::Error>`) an IoError to custom errors? That would allow us to construct custom errors like before using custom contexts, with the content of the IoError appended, right?

As far as I could gather that's pretty much exactly what the current implementation of `UIoError` does anyways.

> You could look at the code coverage results to see if there are any untested errors and add them to the tests if possible. It's not necessary for this PR, though.

Is there a way I can run the coverage tests locally? It feels pretty "wrong" to abuse the CI pipeline for this sort of thing, especially as it isn't blazingly fast.

**Comment by tertsdiepraam** [comment]
> But if multiple utils fail to get Metadata, isn't that a candidate for a `CommonError` then?

Yes, although some are really specific to some utils.

> Maybe one could "attach" (perhaps by means of another attribute that's a Option<std::io::Error>) an IoError to custom errors?

You already can to some extent. In the `Display impl` for the `UCustomError`, you could take the `io::Error` you passed to it, apply context to it and display that. This doesn't solve the boilerplate in the rest of the code though (still requires manual matching on the error and the construction of the custom error).

**Comment by tertsdiepraam** [comment]
As a sidenote, feel free to experiment with changing some of the error things in uucore (maybe in a separate PR?). I definitely do not consider it finished yet and welcome any suggestions :)

**Comment by 353fc443** [comment]
> Is there a way I can run the coverage tests locally? It feels pretty "wrong" to abuse the CI pipeline for this sort of thing, especially as it isn't blazingly fast.

You could run the coverage tests on your fork

**Comment by tertsdiepraam** [comment]
I've been thinking about this some more and I think going with something simple is the best solution. That would be to simply provide a function for printing the IO error messages. Something like:
```rust
// This doesn't exist right now, but we can easily break it out of the UIoError Display impl
use uucore::error::io_error_message;

enum UtilError {
    SomeIOErrorWrapper(PathBuf, io::Error),
    // other variants
}

impl Display for UtilError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            UtilError::SomeIOErrorWrapper(file, e) => {
                write!(f, "some desciption {}: {}", file.display(), io_error_message(e))
            }
            // other variants
        }
    }
}

// Usage:
fs::metadata(&path).map_err(|e| UtilError::SomeIOErrorWrapper(path.into_path_buf(), e))

**Comment by Funky185540** [comment]
Any if we use a macro... ? Is that something a rust macro could do? Something like `uio_error!(e)`, which then converts the `std::io::Error` from my code to a `UIoError` with an enmpty context. Then the call to `write!` would use the `Display` impl from the `UIoError`, right?

**Comment by Funky185540** [comment]
This kinda does the trick:

```rust
#[macro_export]
macro_rules! to_uioerr {
    ( $arg:expr ) => {
        UIoError::new(
            $arg.kind(), String::from("")
        )
    };
}
```

but the `Display` impl always prepends a ": " to the actual error message, which is a bit irritating. This would turn the following error variant:

```rust
            IE::InstallFailed(from, to, e) => write!(
                f,
                "cannot install '{}' to '{}': {}",
                from.display(),
                to.display(),
                e
            ),
```

into:

```rust
            IE::InstallFailed(from, to, e) => write!(
                f,
                "cannot install '{}' to '{}'{}",
                from.display(),
                to.display(),
                to_uioerr!(e)
            ),
```

which looks weird, because in the format string we now have "'{}'{}". :-(

Using the function approach we could have the function return only the error string, and then make the `Display` impl prepend the ": " to that, if desired.

**Comment by Funky185540** [comment]
Or we add this to the `error` module:

```rust
impl From<&std::io::Error> for UIoError {
    fn from(e: &std::io::Error) -> Self {
        UIoError {
            context: String::from(""),
            inner: std::io::Error::new(e.kind(), ""),
        }
    }
}
```

which allows us to write this:

```rust
            IE::InstallFailed(from, to, e) => write!(
                f,
                "cannot install '{}' to '{}': {}",
                from.display(),
                to.display(),
                UIoError::from(e)
            ),
```

I'm fine with either of these: The separate function, the macro or the `From` impl. The separate function sounds most reasonable as this allows us to easily handle the omission of ": " when used like discussed here. 

But if we add that, it sounds as if it deserves its own PR. @tertsdiepraam Do you want to take care of this?

**Comment by tertsdiepraam** [comment]
Very nice to see so many ideas! That really helps with exploring all possibilities. I think the macro doesn't really provide anything over a function, while also being a little bit harder to understand (because macros just in general are a little bit more complicated in my opinion).

Making the context optional is definitely an interesting idea. I wonder if there are other applications for context-less `UIoErrors`. I don't really have a strong preference for this or the function right now.

For reference to the entire discussion, I feel it makes sense to explicitly state the current situation, which isn't too bad I think. I may not be very intuitive though:
```rust
IE::InstallFailed(from, to, e) => {
    e.map_err_context(|| format!(
        "cannot install '{}' to '{}'",
        from.display(),
        to.display()
    )).fmt(f)
},
```

**Comment by Funky185540** [comment]
That's a pretty syntax, I didn't know that was an option!

Just 2 remarks by rust about that:

1. `fmt` here is ambiguous, it could be `Debug::fmt` or `Display::fmt`
2. this call would consume (move) `e` as `std::io::Error` doesn't implement the `Copy` trait:
    ```
    error[E0507]: cannot move out of `*e` which is behind a shared reference
       --> src/uu/install/src/install.rs:115:18
        |
    115 |                 &e.map_err_context(|| format!(
        |                  ^ move occurs because `*e` has type `std::io::Error`, which does not implement the `Copy` trait

    ```

Now I have quite a lot of options I must say. I'd go with your last suggestions, @tertsdiepraam, but I'm unsure what to do about that second error. :-/

**Comment by tertsdiepraam** [comment]
Ah my bad, should've let the compiler check that code. Can you get around that error by cloning `e`?

**Comment by Funky185540** [comment]
Don't worry.

I have now tried this code:

```rust
            IE::InstallFailed(from, to, e) => Display::fmt(
                &e.clone().map_err_context(|| format!(
                    "cannot install '{}' to '{}'",
                    from.display(),
                    to.display()
                )), f
            ),
```

which gives me this pretty error message:

```
error[E0507]: cannot move out of a shared reference
   --> src/uu/install/src/install.rs:115:18
    |
115 |                 &e.clone().map_err_context(|| format!(
    |                  ^^^^^^^^^ move occurs because value has type `std::io::Error`, which does not implement the `Copy` trait
```

And I can't make any sense of it. I *know* I'm moving it, that's exactly the point of the whole thing.

**Comment by tertsdiepraam** [comment]
The error is indeed extremely weird. I *think* it's due to `io:Error` not implementing `clone`. Here's what I think is going on: the `clone` call acts on the reference, which makes a new reference and therefore doesn't solve the issue. This is because `clone` is implemented for all references and not for `std::io::Error`. You can work around this by constructing the `UIoError` from the `ErrorKind` instead, because that implements `Copy`:
```rust
IE::InstallFailed(from, to, e) => Display::fmt(
    &UIoError::new(
        e.kind(),
        format!("cannot install '{}' to '{}'", from.display(), to.display()),
    ),
    f,
),
```

**Comment by Funky185540** [comment]
You're right, that does the trick!

I understand that you're not particularly fond of macros, but something like this:

```rust
#[macro_export]
macro_rules! uioerror(
    ($err:expr, $($args:tt)+) => ({
        UIoError::new(
            $err.kind(),
            format!($($args)+)
        )
    })
);
```

allows us to turn this:

```rust
            IE::InstallFailed(from, to, e) => Display::fmt(
                &UIoError::new(
                    e.kind(), 
                    format!("cannot install '{}' to '{}'", from.display(), to.display()
                )), f
            ),
```

into this:

```rust
            IE::InstallFailed(from, to, e) => Display::fmt(
                &uioerror!(e, "cannot install '{}' to '{}'", from.display(), to.display()), f
            ),
```

I'm fine with either implementation, but I'd be curious to hear some more opinions by others that have faced a similar situation already. I'll leave the decision on which implementation to use up to you. :-)

**Comment by tertsdiepraam** [comment]
I like that macro! I've already been thinking about adding something like that for `USimpleError` and `UUsageError` and it makes a lot of sense here.

**Comment by Funky185540** [comment]
The changes are in, feedback is welcome!

Looking at the `macros` submodule I might spend some time on documenting all the macros in there next... We'll see.

**Comment by tertsdiepraam** [comment]
Looks great! I'm just wondering where that macro should live. (In fact having a macro module feels off to me.) I think that it makes sense for discoverability to put it in the `error` module. Do you agree? (if you don't just say so and I'll merge it like this :) )

Side note: I love the effort you put in documentation in all your PR's. Absolutely fantastic work there!

**Comment by miDeb** [comment]
Bikeshed: Not sure if this is only me, but `uioerror` looks a bit unreadable to me :sweat_smile:. Maybe it can just be `io_error`?

**Comment by Funky185540** [comment]
> I think that it makes sense for discoverability to put it in the error module. Do you agree?

I certainly do! I've moved it there now and edited the commit accordingly.

> Bikeshed: Not sure if this is only me, but uioerror looks a bit unreadable to me

Agreed. But I wouldn't call it `io_error`, because I'd expect that to somehow create `std::io::Error` instances which isn't what it does. I've named it `uio_error` instead, is that okay with you as well?

**Comment by Funky185540** [comment]
Thank god there's this CI pipeline, I always forget either satisfying clippy, cspell or the uucore-specific tests. ;-)

> Side note: I love the effort you put in documentation in all your PR's. Absolutely fantastic work there!

Thanks, I'm glad to hear it!

**Comment by Funky185540** [comment]
@tertsdiepraam This is good to go from my side. What do you think?

File: src/uu/install/src/install.rs [hunk 1]
```diff
@@ -408,98 +488,85 @@ fn standard(mut paths: Vec<String>, b: Behavior) -> i32 {
...
 /// _target_ must be a non-directory
 ///
-fn copy_file_to_file(file: &Path, target: &Path, b: &Behavior) -> i32 {
-    if copy(file, target, b).is_err() {
-        1
-    } else {
-        0
-    }
+fn copy_file_to_file(file: &Path, target: &Path, b: &Behavior) -> UResult<()> {
+    copy(file, target, b)
```
Comment (tertsdiepraam) [inline src/uu/install/src/install.rs:563]: I think this function is now useless :)

File: src/uu/install/src/install.rs [hunk 2]
```diff
@@ -408,98 +488,85 @@ fn standard(mut paths: Vec<String>, b: Behavior) -> i32 {
...
         }
 
         let mut targetpath = target_dir.to_path_buf();
         let filename = sourcepath.components().last().unwrap();
         targetpath.push(filename);
 
-        if copy(sourcepath, &targetpath, b).is_err() {
-            all_successful = false;
+        if let Err(e) = copy(sourcepath, &targetpath, b) {
+            show!(e);
```
Comment (tertsdiepraam) [inline src/uu/install/src/install.rs:543]: Not necessary, but there is a `show_if_err!` macro for this purpose. If you think this is clearer, feel free to leave it in

**Review by tertsdiepraam — COMMENTED** [review]
Almost perfect! Just 1 function that's no longer needed because of the changes you made.

File: src/uu/install/src/install.rs (hunk 1 above)
Comment (Funky185540) [reply src/uu/install/src/install.rs]: Well spotted, I completely missed that one. Thanks!

File: src/uu/install/src/install.rs (hunk 2 above)
Comment (Funky185540) [reply src/uu/install/src/install.rs]: You're absolutely right! It's much nicer to read that way, too. Added it.

**Comment by Funky185540** [comment]
Fascinating. The CI is failing on a line of code that seemingly doesn't exist...

**Comment by tertsdiepraam** [comment]
Hmm yeah, I'm not sure what that's about. Rerunning the CI doesn't help. I did find this snippet:
```rust
    let specified_mode: Option<u32> = if matches.is_present(OPT_MODE) {
        let x = matches.value_of(OPT_MODE).ok_or(1)?;
        Some(mode::parse(x, considering_dir).map_err(|err| {
            show_error!("Invalid mode string: {}", err);
            1
        })?)
    } else {
        None
    };
```

Which returns `Err(1)` I think, maybe that's what it's complaining about? Or something similar with `backup_control`?

**Comment by tertsdiepraam** [comment]
Maybe you could merge master into this just to make some change that will rerun the CI?

**Comment by Funky185540** [comment]
Rebased onto master, now I see it, too.

Hold on. That means that Git implicitly rebases my branch onto master when running the CI stuff. Is this intentional?!

**Comment by tertsdiepraam** [comment]
> That means that Git implicitly rebases my branch onto master when running the CI stuff. Is this intentional?

I think so. In practice we don't really care whether a specific PR succeeds in the tests, but rather whether master succeeds when the PR is merged, so this makes sense (although it may lead to confusing situations)

**Comment by Funky185540** [comment]
That of course makes a lot of sense. :)

It's sad that the coverage is broken... but from my side this is done now.

**Comment by sylvestre** [comment]
I pinged grcov author yesterday and i have a meeting with him later today ;) I will bother him about that! cc @marco-c

**Comment by miDeb** [comment]
Can't we just use a newer rust version for coverage? [Currently it's set to `nightly-2020-08-01`](https://github.com/uutils/coreutils/blob/978033a241d0204575eaf3141614809513c44873/.github/workflows/CICD.yml#L17), maybe we can change this to the latest stable release?

File: src/uu/install/src/install.rs [hunk 3]
```diff
@@ -5,7 +5,10 @@
 //  * For the full copyright and license information, please view the LICENSE file
 //  * that was distributed with this source code.
 
-// spell-checker:ignore (ToDO) rwxr sourcepath targetpath
+// spell-checker:ignore (ToDO) rwxr sourcepath targetpath Isnt uioerror
+
+// clippy bug https://github.com/rust-lang/rust-clippy/issues/7422
+#![allow(clippy::nonstandard_macro_braces)]
```
Comment (miDeb) [inline src/uu/install/src/install.rs:11]: You can remove these lines again, the clippy bug has been fixed.

Comment (Funky185540) [reply src/uu/install/src/install.rs]: Done!

**Comment by sylvestre** [comment]
@Funky185540 could you please rebase it? it will trigger the code coverage. (i fixed it)

**Comment by Funky185540** [comment]
Rebased onto current master.

**Pull request merged**